### PR TITLE
WIP : python3-ization and type hinting

### PR DESCRIPTION
**Pull request by hadware**
Same thing as for pyannote-core.

First question: for what you call "array-like", i'm proposing 3 solutions:

 - create this type and use it everywhere : `ArrayLike = Union[Sequence, np.ndarray]`
 - only use `Sequence`
 - only use `np.ndarray` 

As for now, it doesn't look like there is anything usable for type-hinting array sizes (after a quick research). 
Ref: https://mypy.readthedocs.io/en/latest/cheat_sheet_py3.html#standard-duck-types (for Sequence)

### CODE REVIEW

**Comment by hbredin** [comment]
Happy new year @hadware !

Thanks for this pull request and sorry for the delay in looking at it.
Looks like it does not pass the tests. Can you please fix that when you find the time?

**Comment by hadware** [comment]
Yes I will! I'll be here on next tuesday @ENS by the way.

**Comment by hadware** [comment]
Well, this should be close to done.

Regarding my former question: there's a nifty `ArrayLike` type in `numpy.typing` that does the job nicely.

File: pyannote/metrics/binary_classification.py [hunk 1]
```diff
@@ -26,15 +26,21 @@
 # AUTHORS
 # Hervé BREDIN - http://herve.niderb.fr
 
-import numpy as np
 from collections import Counter
+from typing import Tuple
+
+import numpy as np
 import sklearn.metrics
+from pandas._typing import ArrayLike
```
Comment (hbredin) [inline pyannote/metrics/binary_classification.py:34]: Why not from numpy?

**Comment by hadware** [comment]
Fixed your comments. I'm wondering: in some cases (notably: for all detection metrics), the required type for ref/hypothesis seems to be `Annotation` (because it then calls `uemify` which "tacitly" only accepts `Annotation` instances). Shouldn't uemify be able to also process `Timeline` instances?

Moreover, maybe we could extend the use of `collar` and `skip_overlap` (used in IER) to other metrics?

**Comment by hbredin** [comment]
> Fixed your comments. I'm wondering: in some cases (notably: for all detection metrics), the required type for ref/hypothesis seems to be `Annotation` (because it then calls `uemify` which "tacitly" only accepts `Annotation` instances). Shouldn't uemify be able to also process `Timeline` instances?

It could be applied to `Timeline` instances but only do the cropping part, but not the common timeline part.
This is because, `Timeline` instances are mostly used for segmentation metrics and we don't want to artificially split long segments into shorter ones.

> Moreover, maybe we could extend the use of `collar` and `skip_overlap` (used in IER) to other metrics?

Once again, we'd have to be careful about segmentation metrics as these two notions probably don't make sense here.

Anyway, this should not be part of this PR.

Shall I merged this PR or do you plan to add more typing?

**Comment by hadware** [comment]
Ok, maybe in another PR, it'll indeed take some time to look into it.

Regarding this current PR, I think i'm mostly done with typing, i'll re-check everything, and it should be good.

**Comment by hbredin** [comment]
C'est votre dernier mot, Jean-Pierre ?

**Comment by hadware** [comment]
image

File: pyannote/metrics/spotting.py [hunk 2]
```diff
@@ -63,15 +74,20 @@ class LowLatencySpeakerSpotting(BaseMetric):
...
 
     @classmethod
-    def metric_name(cls):
+    def metric_name(cls) -> str:
         return "Low-latency speaker spotting"
 
-    @classmethod
-    def metric_components(cls):
-        return {'target': 0.}
+    def metric_components(self) -> MetricComponents:
```
Comment (hbredin) [inline pyannote/metrics/spotting.py:80]: Not quite sure about this change.

Components are meant to be accumulated, which is not the case for those "components" you added.

Comment (hbredin) [reply pyannote/metrics/spotting.py:80]: Not all. Only those related to components.

**Comment by hbredin** [comment]
🎉 Thanks @hadware !

**Pull request merged**